File: internal/auto-commit/auto-commit_test.go
package autocommit

import (
	"os"
	"os/exec"
	"path/filepath"
	"testing"
)

func setupTestGitRepo(t *testing.T) string {
	dir := t.TempDir()

	// Create required directories
	err := os.MkdirAll(filepath.Join(dir, "wiki"), 0755)
	if err != nil {
		t.Fatal(err)
	}
	err = os.MkdirAll(filepath.Join(dir, "state"), 0755)
	if err != nil {
		t.Fatal(err)
	}

	cmd := exec.Command("git", "init")
	cmd.Dir = dir
	if err := cmd.Run(); err != nil {
		t.Fatal(err)
	}

	cmd = exec.Command("git", "config", "user.email", "test@example.com")
	cmd.Dir = dir
	if err := cmd.Run(); err != nil {
		t.Fatal(err)
	}

	cmd = exec.Command("git", "config", "user.name", "Test User")
	cmd.Dir = dir
	if err := cmd.Run(); err != nil {
		t.Fatal(err)
	}

	err = os.WriteFile(filepath.Join(dir, "wiki", "init.md"), []byte("init"), 0644)
	if err != nil {
		t.Fatal(err)
	}

	cmd = exec.Command("git", "add", ".")
	cmd.Dir = dir
	if err := cmd.Run(); err != nil {
		t.Fatal(err)
	}

	cmd = exec.Command("git", "commit", "-m", "init")
	cmd.Dir = dir
	if err := cmd.Run(); err != nil {
		t.Fatal(err)
	}

	return dir
}

func TestIsGitRepo(t *testing.T) {
	t.Run("returns true for git repo", func(t *testing.T) {
		dir := setupTestGitRepo(t)
		if !IsGitRepo(dir) {
			t.Errorf("Expected IsGitRepo to return true")
		}
	})

	t.Run("returns false for non-git repo", func(t *testing.T) {
		dir := t.TempDir()
		if IsGitRepo(dir) {
			t.Errorf("Expected IsGitRepo to return false")
		}
	})
}

func TestAutoCommitWikiChanges(t *testing.T) {
	t.Run("skips when autoCommit is false and no force", func(t *testing.T) {
		dir := setupTestGitRepo(t)

		config := VaultConfigPayload{
			AutoCommit: false,
			WikiDir:    filepath.Join(dir, "wiki"),
			StateDir:   filepath.Join(dir, "state"),
		}

		msg, err := AutoCommitWikiChanges(dir, "test", "detail", config, AutoCommitOptions{})
		if err != nil {
			t.Fatal(err)
		}
		if msg != nil {
			t.Errorf("Expected msg to be nil, got %v", *msg)
		}
	})

	t.Run("commits when forced even if autoCommit is false", func(t *testing.T) {
		dir := setupTestGitRepo(t)

		// Create a file to commit
		err := os.WriteFile(filepath.Join(dir, "wiki", "test.md"), []byte("test content"), 0644)
		if err != nil {
			t.Fatal(err)
		}

		config := VaultConfigPayload{
			AutoCommit: false,
			WikiDir:    filepath.Join(dir, "wiki"),
			StateDir:   filepath.Join(dir, "state"),
		}

		msg, err := AutoCommitWikiChanges(dir, "test", "detail", config, AutoCommitOptions{Force: true})
		if err != nil {
			t.Fatal(err)
		}

		if msg == nil {
			t.Fatal("Expected commit message, got nil")
		}
		expected := "vault test: detail"
		if *msg != expected {
			t.Errorf("Expected message %q, got %q", expected, *msg)
		}
	})

	t.Run("commits when autoCommit is true", func(t *testing.T) {
		dir := setupTestGitRepo(t)

		// Create a file to commit
		err := os.WriteFile(filepath.Join(dir, "wiki", "test.md"), []byte("test content"), 0644)
		if err != nil {
			t.Fatal(err)
		}

		config := VaultConfigPayload{
			AutoCommit: true,
			WikiDir:    filepath.Join(dir, "wiki"),
			StateDir:   filepath.Join(dir, "state"),
		}

		msg, err := AutoCommitWikiChanges(dir, "test", "detail", config, AutoCommitOptions{})
		if err != nil {
			t.Fatal(err)
		}

		if msg == nil {
			t.Fatal("Expected commit message, got nil")
		}
		expected := "vault test: detail"
		if *msg != expected {
			t.Errorf("Expected message %q, got %q", expected, *msg)
		}
	})

	t.Run("skips when no changes to commit", func(t *testing.T) {
		dir := setupTestGitRepo(t)

		config := VaultConfigPayload{
			AutoCommit: true,
			WikiDir:    filepath.Join(dir, "wiki"),
			StateDir:   filepath.Join(dir, "state"),
		}

		msg, err := AutoCommitWikiChanges(dir, "test", "detail", config, AutoCommitOptions{})
		if err != nil {
			t.Fatal(err)
		}

		if msg != nil {
			t.Errorf("Expected msg to be nil, got %v", *msg)
		}
	})
}

File: packages/engine/src/auto-commit.test.ts
import { execFile } from "node:child_process";
import { mkdir, mkdtemp, rm, writeFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { promisify } from "node:util";
import { afterEach, beforeEach, describe, expect, it } from "vitest";
import { autoCommitWikiChanges } from "./auto-commit.js";

const execFileAsync = promisify(execFile);

async function git(rootDir: string, ...args: string[]): Promise<string> {
  const { stdout } = await execFileAsync("git", args, { cwd: rootDir });
  return stdout.trim();
}

async function setupTestGitRepo(): Promise<string> {
  const dir = await mkdtemp(join(tmpdir(), "swarmvault-test-"));
  await mkdir(join(dir, "wiki"), { recursive: true });
  await mkdir(join(dir, "state"), { recursive: true });

  const defaultConfigStr = JSON.stringify({
    providers: {
      local: {
        type: "heuristic",
        model: "heuristic-v1",
        capabilities: ["chat", "structured", "vision", "local"]
      }
    },
    tasks: {
      compileProvider: "local",
      queryProvider: "local",
      lintProvider: "local",
      visionProvider: "local",
      imageProvider: "local"
    }
  });

  await writeFile(join(dir, "swarmvault.config.json"), JSON.stringify({ autoCommit: true, ...JSON.parse(defaultConfigStr) }));

  await git(dir, "init");
  await git(dir, "config", "user.email", "test@example.com");
  await git(dir, "config", "user.name", "Test User");

  await writeFile(join(dir, "wiki", "init.md"), "init");
  await git(dir, "add", ".");
  await git(dir, "commit", "-m", "init");

  return dir;
}

const runTests = () => {
  it("skips when autoCommit is false and no force", async () => {
    const dir = await setupTestGitRepo();
    const defaultConfigStr = JSON.stringify({
      providers: {
        local: {
          type: "heuristic",
          model: "heuristic-v1",
          capabilities: ["chat", "structured", "vision", "local"]
        }
      },
      tasks: {
        compileProvider: "local",
        queryProvider: "local",
        lintProvider: "local",
        visionProvider: "local",
        imageProvider: "local"
      }
    });
    await writeFile(join(dir, "swarmvault.config.json"), JSON.stringify({ autoCommit: false, ...JSON.parse(defaultConfigStr) }));
    const msg = await autoCommitWikiChanges(dir, "test", "detail", {});
    expect(msg).toBeNull();
    await rm(dir, { recursive: true, force: true });
  });

  it("commits when forced even if autoCommit is false", async () => {
    const dir = await setupTestGitRepo();
    const defaultConfigStr = JSON.stringify({
      providers: {
        local: {
          type: "heuristic",
          model: "heuristic-v1",
          capabilities: ["chat", "structured", "vision", "local"]
        }
      },
      tasks: {
        compileProvider: "local",
        queryProvider: "local",
        lintProvider: "local",
        visionProvider: "local",
        imageProvider: "local"
      }
    });
    await writeFile(join(dir, "swarmvault.config.json"), JSON.stringify({ autoCommit: false, ...JSON.parse(defaultConfigStr) }));
    await writeFile(join(dir, "wiki", "test.md"), "test content");

    const msg = await autoCommitWikiChanges(dir, "test", "detail", { force: true });
    expect(msg).toBe("vault test: detail");
    await rm(dir, { recursive: true, force: true });
  });

  it("commits when autoCommit is true", async () => {
    const dir = await setupTestGitRepo();
    await writeFile(join(dir, "wiki", "test.md"), "test content");

    const msg = await autoCommitWikiChanges(dir, "test", "detail", {});
    expect(msg).toBe("vault test: detail");
    await rm(dir, { recursive: true, force: true });
  });

  it("skips when no changes to commit", async () => {
    const dir = await setupTestGitRepo();

    const msg = await autoCommitWikiChanges(dir, "test", "detail", {});
    expect(msg).toBeNull();
    await rm(dir, { recursive: true, force: true });
  });
};

describe("autoCommitWikiChanges (TS Mode)", () => {
  beforeEach(() => {
    delete process.env.USE_GO_PORT;
  });
  runTests();
});

describe("autoCommitWikiChanges (Go Sidecar Mode)", () => {
  beforeEach(() => {
    process.env.USE_GO_PORT = "true";
  });
  afterEach(() => {
    delete process.env.USE_GO_PORT;
  });
  runTests();
});
